pkg/controller/cluster/controller.go

```diff
@@ -23,33 +23,112 @@ import (
+	"github.com/golang/glog"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/cluster-api/util"
+	"k8s.io/client-go/tools/cache"
 )
 
 // +controller:group=cluster,version=v1alpha1,kind=Cluster,resource=clusters
 type ClusterControllerImpl struct {
 	builders.DefaultControllerFns
 
-	// lister indexes properties about Cluster
-	lister listers.ClusterLister
+	actuator Actuator
+
+	// clusterLister holds a lister that knows how to list Cluster from a cache
+	clusterLister listers.ClusterLister
+	// machineLister holds a lister that knows how to list Machines from a cache
+	machineLister listers.MachineLister
+
+	// machineListerSynced function whether the machine lister is synced
+	clusterListerSynced cache.InformerSynced
+	// machineListerSynced function whether the machine lister is synced
+	machineListerSynced cache.InformerSynced
+
+	kubernetesClientSet *kubernetes.Clientset
+	clientSet           clientset.Interface
+	clusterClient       v1alpha1.ClusterInterface
 }
 
 // Init initializes the controller and is called by the generated code
 // Register watches for additional resource types here.
-func (c *ClusterControllerImpl) Init(arguments sharedinformers.ControllerInitArguments) {
-	// Use the lister for indexing clusters labels
-	c.lister = arguments.GetSharedInformers().Factory.Cluster().V1alpha1().Clusters().Lister()
+func (c *ClusterControllerImpl) Init(arguments sharedinformers.ControllerInitArguments, actuator Actuator) {
+	cInformer := arguments.GetSharedInformers().Factory.Cluster().V1alpha1().Clusters()
+	mInformer := arguments.GetSharedInformers().Factory.Cluster().V1alpha1().Machines()
+
+	c.clusterLister = cInformer.Lister()
+	c.machineLister = mInformer.Lister()
+
+
+	c.actuator = actuator
+
+	c.clusterListerSynced = cInformer.Informer().HasSynced
+	c.machineListerSynced = mInformer.Informer().HasSynced
+
+	cs, err := clientset.NewForConfig(arguments.GetRestConfig())
+	if err != nil {
+		glog.Fatalf("error creating cluster client: %v", err)
+	}
+	c.clientSet = cs
+	c.kubernetesClientSet = arguments.GetSharedInformers().KubernetesClientSet
+
+	// Create cluster actuator.
+	// TODO: Assume default namespace for now. Maybe a separate a controller per namespace?
+	c.clusterClient = cs.ClusterV1alpha1().Clusters(corev1.NamespaceDefault)
+	c.actuator = actuator
 }
 
-// Reconcile handles enqueued messages
-func (c *ClusterControllerImpl) Reconcile(u *v1alpha1.Cluster) error {
-	// Implement controller logic here
-	log.Printf("Running reconcile Cluster for %s\n", u.Name)
-	return nil
+func (c *ClusterControllerImpl) Run(stopCh <-chan struct{}) {
+	glog.Infof("Waiting for caches to sync for cluster controller")
+	if !cache.WaitForCacheSync(stopCh, c.machineListerSynced, c.clusterListerSynced) {
+		glog.Warningf("Unable to sync caches for cluster controller")
+		return
+	}
+	glog.Infof("Caches are synced for machineset controller")
 }
```

Comment:
(sanity check)
I was under the impression that Run should run until it is sopped by the stopCh. This seems to return after the cache syncs. Am I missing something?

Reply:
I modeled this after the machineset [PR](https://github.com/kubernetes-sigs/cluster-api/pull/186/files#diff-b0f93c81bd86491a95181a762269c619R93). @k4leung4 do you know if this is supposed to continuously sync the cache until it is stopped? There is another run method [here](https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/cluster/zz_generated.api.register.go#L119) that actually starts a bunch of worker queues which are stopped by stopCh.

Reply:
There are multiple runs and this is due to how the apiserver-builder generated the code.

In the zz_generated file, there it has the Run for the base Controller obj, which calls Run on the workers which does run until it is stopped by the stopCh. After the workers Run are kicked off, it will attempt to call Run on the ControllerImpl obj if one is defined.

So @jessicaochen, you are right about the behavior but Run is overloaded as it is sharded into three different places, the workers, the base controller and controller impl.

Reply:
@k4leung4 can you link to an example of the generated code that calls this Run(...) method? It would be helpful in reasoning about the correct behavior of this function.

Reply:
This is generated/base controller Run that calls the controller impl Run
https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machineset/zz_generated.api.register.go#L119

The worker Run
https://github.com/kubernetes-sigs/cluster-api/blob/master/vendor/github.com/kubernetes-incubator/apiserver-builder/pkg/controller/controller.go#L79

Reply:
Hello @k4leung4 thanks for the deep dive.

Reply:
I removed the WaitForCacheSync call from this PR